text/0000-test-framework-binnacle.md

```diff
@@ -0,0 +1,357 @@
+        "device": "/dev/vdc"
+    },
+    {
+        "node": "node2.example.com",
+        "device": "/dev/vdc"
+    },
+    {
+        "node": "node3.example.com",
+        "device": "/dev/vdc"
+    }
+]
+"""
+
+LOOP:
+    NODE=$node
+    TEST mkfs.xfs $device
+```
+
+In case of external JSON data,
+
+```
+LOOP_DATA_FILE=device_config.json
+LOOP:
+    NODE=$node
+    TEST mkfs.xfs $device
+```
+
+### Ease of use
+
+No special syntax. The example below is to test the Gluster Volume
+force start command.
+
+```
+N1=node1.example.com
+N2=node2.example.com
+N3=node3.example.com
+volname=gv1
+
+NODE=$N1
+TEST gluster volume create ${volname} \
+    $N1:/exports/bricks/${volname}/brick1/brick \
+    $N2:/exports/bricks/${volname}/brick2/brick
+TEST gluster volume start ${volname}
+TEST_WITHIN -t 60 gluster volume status ${volname} | match_num_bricks_online 2
+
+NODE=$N2
+TEST brick_kill "/exports/bricks/${volname}/brick2/brick"
+
+NODE=$N1
+TEST_WITHIN -t 60 gluster volume status ${volname} | match_num_bricks_online 1
+TEST gluster volume start ${volname} force
+TEST_WITHIN -t 60 gluster volume status ${volname} | match_num_bricks_online 2
+```
+
+### Kadalu Test example
+
+```
+MASTER=master.example.com
+N1=kube-node1.example.com
+N2=kube-node2.example.com
+N3=kube-node3.example.com
+STORAGE_POOL_NAME=sp1
+OPERATOR_YAML=operator.yaml
+PVC_NAME=pv1
+PVC_FILE_NAME=pv1.yaml
+APP_POD_NAME=pod1
+APP_POD_FILE=pod1.yaml
+
+NODE=$MASTER
+TEST kube_master_setup --record-output tmp/master-details.json
+
+LOOP_DATA="""
+[
+  {
+    "node": "$N1",
+    "device": "/dev/vdc"
+  },
+  {
+    "node": "$N2",
+    "device": "/dev/vdc"
+  },
+  {
+    "node": "$N3",
+    "device": "/dev/vdc"
+  }
+]
+"""
+LOOP:
+    NODE=$node
+    TEST kube_node_setup
+    TEST join_kube_master tmp/master-details.json
+    TEST loop_device_setup $device
+
+NODE=$MASTER
+TEST kubectl create -f $OPERATOR_YAML
+EXPECT_WITHIN -t 180 -v 3 kubectl get pods -n kadalu | grep "Running" | wc -l
+TEST kubectl kadalu storage-add $STORAGE_POOL_NAME --type Replica3 \
+    --device $N1:/dev/vdc
+    --device $N2:/dev/vdc
+    --device $N3:/dev/vdc
+EXPECT_WITHIN -t 180 -v 3 kubectl get pods -n kadalu | grep $STORAGE_POOL_NAME | wc -l
+
+# Sample PVC Test
+TEST kubectl create -f $PVC_FILE
+EXPECT_WITHIN -t 60 -v 1 kubectl get pvc | grep $PVC_NAME | grep Bound | wc -l
+
+# Sample app Test
+TEST kubectl create -f $APP_POD_FILE
+EXPECT_WITHIN -t 180 -v 1 kubectl get pods | grep $APP_POD_NAME | grep Running | wc -l
+
+```
+
+## How it works?
+
+Similar to Gluster upstream tests, Binnacle uses [Test Anything
+Protocol(TAP)](https://en.wikipedia.org/wiki/Test_Anything_Protocol). This
+test framework provides syntax sugar on top of the shell script.
+
+Binnacle generates shell script from the given test file and then
+executes the generated script using
+[prove](https://metacpan.org/pod/distribution/Test-Harness/bin/prove)
+command. The following needs to be changed to bash before actually
```

Comment:
following? Couldn't get what following needs to be changed.

Reply:
Removed